### PR TITLE
feat(schemas): add splitByTags option to organize schemas by tag

### LINKED ISSUE #3592: feat: support tags-split mode for schemas output
## Problem

When using `output.mode: 'tags-split'` with a separate `schemas.path`, client service files are organized into per-tag subdirectories, but all schema/model files are written flat into a single directory. For large APIs this creates a flat directory with hundreds of files, making navigation difficult and changes hard to review.

Example layout:
```
client/sdk/generated/             # per-tag clients ✓
  pet/pet.service.ts
  store/store.service.ts

models/sdk/generated/             # flat schemas ✗
  pet.ts
  store.ts
  order.ts              # shared across tags
  getApiPetById...ts    # HTTP param types
  ... hundreds more
```

## Breakdown

For a typical large OpenAPI spec, generated schema files fall into three categories:
- **Tag-prefixed schemas** — the majority, clearly belonging to a single tag (e.g. `pet...`, `store...`)
- **HTTP param types** — operation-specific request/response param types (`getApi...`, `postApi...`, etc.)
- **Shared/generic schemas** — types referenced by multiple tags (e.g. pagination wrappers, error types, common identifiers)

## Suggested API

Add a `schemas.mode` option that mirrors schemas into tag-based subdirectories:

```ts
output: {
  mode: 'tags-split',
  schemas: {
    path: './models/sdk/generated',
    importPath: '@acme/models',
    mode: 'tags-split',
  },
}
```

Would produce:

```
models/sdk/generated/
  pet/
    pet.ts
    petNotFoundError.ts
    index.ts
  store/
    store.ts
    storeInventory.ts
    index.ts
  _shared/                    # schemas used by 2+ tags
    pagination.ts
    error.ts
    index.ts
  index.ts                    # root barrel re-exporting all
```

## Shared schema handling

Schemas referenced by multiple tags are the main design concern. Options:

1. **Shared bucket** — schemas used by 2+ tags go in `_shared/` (or `common/`). Tag-specific schemas go in their tag dir. This is explicit and easy to understand.
2. **Primary tag** — place in the first tag that references it. Fragile and implicit — tag changes can silently move files.
3. **All flat (current)** — simple but defeats the purpose of splitting.

I suggest option 1 (shared bucket) with a configurable directory name (default: `_shared`).

## Notes

- A root `index.ts` barrel re-exporting all subdirectories would preserve the current flat import experience (`import { Pagination } from '@acme/models'`).
- Cross-tag schema references would use relative imports within generated files (e.g. `import { Pagination } from '../_shared/pagination'`).
- HTTP param types (`getApi...`, `postApi...`) would go in the tag directory matching their operation's tag.

### PR DESCRIPTION
**Pull request by aqeelat**
## Summary

Adds `schemas.splitByTags` option that organizes generated schema files into per-tag subdirectories. Schemas referenced by multiple tags remain at the root of the schema directory. Works with any `output.mode` (`single`, `split`, `tags`, `tags-split`), both TypeScript and zod schemas, and both `indexFiles: true` (default) and `indexFiles: false`.

Closes #3592

## Configuration

```yaml
output:
  mode: tags-split
  target: ./endpoints.ts
  schemas:
    path: ./model
    splitByTags: true
```

## Generated layout

With `indexFiles: true` (default):

```
model/
  error.ts            # shared schemas at root
  pagination.ts
  pets/
    pet.ts
    createPetsBody.ts
    index.ts
  stores/
    store.ts
    order.ts
    index.ts
  index.ts            # root barrel: export * from "./error", "./pets", "./stores"
```

With `indexFiles: false`, no root or per-tag `index.ts` barrels are emitted. Each operation file and the consolidated `<schemas>/index.faker.ts` import schemas directly from their on-disk location (e.g. `../model/pets/pet`, `./error`). Operation imports are routed via the schema-to-tag map so they resolve into the correct tag subdirectory.

## Implementation

- **`schema-tag-mapper.ts`** — `buildSchemaTagMap()` maps each schema to a tag (or root `.` for shared). Uses operation imports for direct assignment, then a fixpoint loop for transitive propagation. Schemas with 0 or >1 tags stay at root.
- **`schemas-tags-split.ts`** — `writeSchemasTagsSplit()` groups schemas by tag, rewrites cross-tag import paths, delegates per-tag writing to existing `writeSchemas`, writes root barrel that re-exports shared files + tag directory indexes.
- **`write-zod-specs.ts`** — Thread `schemaTagMap` through `writeZodSchemas`, `writeZodSchemasReusable`, and `writeZodSchemasFromVerbs` so zod schemas are organized into per-tag subdirectories. `buildSiblingImports` computes cross-directory paths. `writeZodSchemaTagsSplitBarrel` writes per-tag + root barrel indexes.
- **`generate-imports-for-builder.ts`** — New optional `schemaTagMap` parameter; the `indexFiles: false` branch routes each schema import into its tag subdirectory. Lookup uses `schemaImport.name` (TS identifier) since `buildSchemaTagMap` keys on `schema.name`.
- **`single-mode.ts` / `split-mode.ts` / `tags-mode.ts` / `split-tags-mode.ts`** — Destructure `schemaTagMap` from `WriteModeProps` and forward to all `generateImportsForBuilder` call sites.
- **`writeFakerSchemaMocks`** (`write-specs.ts`) — When `indexFiles: false`, the consolidated `<schemas>/index.faker.ts` routes each schema type import to its on-disk path (`./<tag>/<file>` or `./<file>` for shared) instead of assuming a root barrel.
- **`writeSpecs`** — Computes `schemaTagMap` once at function scope (when `splitByTags` is enabled) and threads it to `writeSchemasTagsSplit`, `writeZodSchemas*`, `writeFakerSchemaMocks`, and the mode-writer dispatch.
- Barrel exports use `./dir` for standard resolution and `./dir/index.js` for NodeNext.

## Validations

- Incompatible with `output.operationSchemas` — operation-derived types are placed within their tag directories automatically.

## Test coverage

- 7 unit tests for `buildSchemaTagMap()` (direct assignment, transitive propagation, multi-tag, shared at root)
- 9 unit tests for `writeSchemasTagsSplit()` (per-tag dirs, shared at root, cross-tag imports, ESM extension, `indexFiles:false`, edge cases, barrel contents)
- 4 integration tests in `generate-spec.test.ts` (works with split mode, rejects with operationSchemas, generates zod schemas, end-to-end generation)
- 2 unit tests in `generate-imports-for-builder.test.ts` for tag-routing behavior (TS identifier vs `schemaName`, fallback to root)
- E2e configs + snapshots:
  - `tagsSplitSharedModels` — multi-tag with shared schemas (pets/stores)
  - `splitModeSplitSchemas` — `mode: 'split'` + `splitByTags`
  - `splitByTagsFakerSchemasNoIndex` — `indexFiles: false` + `faker schemas: true` (the previously-broken combination)
- 5 e2e regression tests in `api-generation.spec.ts` pinning per-tag layout, barrel imports, faker factory imports under `indexFiles:false`, and cross-tag isolation

## Follow-ups

- #3623 — Inconsistent file-existence checks across test files (`readFile` vs `pathExists`)
- #3627 — `single-mode` missing faker schema-factory import recovery (asymmetric with the other three mode writers; pre-existing from #3607, unrelated to this PR)

## Summary by CodeRabbit

* **New Features**
  * Added `splitByTags` configuration option to organize generated schemas into per-tag subdirectories; schemas referenced by multiple tags or no tags remain at the root.

* **Documentation**
  * Documented tag-aware schema directory splitting with configuration examples, directory structure guidance, and import path resolution instructions.

### CODE REVIEW

File: packages/core/src/writers/schema-tag-mapper.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    if (!schemaToTags.has(schema.name)) {
+      schemaToTags.set(schema.name, new Set());
+    }
+  }
+
+  for (const operation of operations) {
+    const tag = kebab(operation.tags[0] ?? DefaultTag);
+    for (const imp of operation.imports) {
+      if (!imp.importPath && schemaNames.has(imp.name)) {
+        addTag(schemaToTags, imp.name, tag);
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/writers/schema-tag-mapper.ts:29]: Investigated this — the suggestion would actually break matching. Here's why:

The tag mapper keys everything off `GeneratorSchema.name`, which is the PascalCase TypeScript identifier (e.g. `PetDto`). The `schemaNames` set and `schemaToTags` map are both built from this:

```ts
const schemaNames = new Set(schemas.map((s) => s.name));  // e.g. { "PetDto" }
```

`GeneratorImport.schemaName` is a **different identifier** — it's the original spec key (e.g. `pet`), set in `resolvers/ref.ts` as `{ name: typeName, schemaName: originalName }`. Using `imp.schemaName ?? imp.name` would substitute `pet` when looking up `PetDto` in the set, causing the match to fail and the schema to be incorrectly bucketed to `.`.

This is also consistent with `normalizeCanonicalImportPaths` in `writers/schemas.ts` (the existing, well-tested canonical import resolver), which also matches purely on `imp.name` — not `schemaName`. The two fields live in different identifier spaces and must not be mixed.

No change made.

**Comment by aqeelat** [comment]
@coderabbitai full review

**Comment by aqeelat** [comment]
@melloware I have the zod support code ready. Should I add it to this PR or in a new PR?

**Comment by melloware** [comment]
You can do the zod support in this PR

**Comment by aqeelat** [comment]
@melloware are you open to making the new flag true by default? Or even removing the flag entirely and making the truthy path the only one. This can be discussed later though.

**Comment by melloware** [comment]
always don't break backwards compatibility first. Then we can discuss switching it later

**Comment by aqeelat** [comment]
@coderabbitai resume

File: packages/orval/src/write-zod-specs.ts [hunk 2]
```diff
@@ -422,6 +481,63 @@ async function writeZodSchemaIndex(
...
+    })
+    .toSorted();
+
+  const tagDirs = [...allDirs.keys()]
+    .filter((dir) => dir !== ROOT_DIR)
+    .toSorted((a, b) => a.localeCompare(b, 'en', { numeric: true }));
+
+  const tagExports = tagDirs.map((dir) => {
+    return `export * from './${dir}';`;
+  });
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/orval/src/write-zod-specs.ts:546]: The suggestion to use `importExt` (`.zod`) for tag dir re-exports would break — per-tag barrel files are `index.ts`, not `index.zod.ts`, so `./pets/index.zod` produces `TS2307: Cannot find module`.

The bare `./pets` works under standard (CommonJS) module resolution. For NodeNext, tag dir re-exports now use `getImportExtension('.ts\, tsconfig)` → `./pets/index.js` (fixed in the latest push).

The schema file exports (`./error.zod`) are a separate pre-existing gap — tracked in #3601.

**Comment by aqeelat** [comment]
@coderabbitai resume
@coderabbitai full review

**Comment by aqeelat** [comment]
@coderabbitai resume
@coderabbitai full review

**Comment by aqeelat** [comment]
@coderabbitai resume
@coderabbitai full review

**Comment by aqeelat** [comment]
@melloware The pr is ready now.

**Comment by melloware** [comment]
@aqeelat is this one ready for review?

**Comment by aqeelat** [comment]
@melloware yes

**Review by melloware — APPROVED** [review]
